src/commands/preview/dev_server.rs

```diff
@@ -0,0 +1,190 @@
+use std::net::{SocketAddr, ToSocketAddrs};
+
+use chrono::prelude::*;
+
+use hyper2::client::{HttpConnector, ResponseFuture};
+use hyper2::header::{HeaderName, HeaderValue};
+use hyper2::service::{make_service_fn, service_fn};
+use hyper2::{Body, Client, Request, Server, Uri};
+
+use hyper_tls2::HttpsConnector;
+
+use failure::format_err;
+
+use uuid::Uuid;
+
+use url::Url;
+
+use crate::commands;
+use crate::commands::preview::upload;
+
+use crate::settings::global_user::GlobalUser;
+use crate::settings::target::Target;
+
+use crate::terminal::emoji;
+
+const PREVIEW_HOST: &str = "rawhttp.cloudflareworkers.com";
+
+#[derive(Clone)]
+struct ServerConfig {
+    host: String,
+    listening_address: SocketAddr,
+    is_https: bool,
+}
+
+impl ServerConfig {
+    pub fn new(
+        host: Option<&str>,
+        ip: Option<&str>,
+        port: Option<&str>,
+    ) -> Result<Self, failure::Error> {
+        let port = port.unwrap_or("8000");
+
+        let try_address = match ip {
+            Some(ip) => format!("{}:{}", ip, port),
+            None => format!("localhost:{}", port),
+        };
+
+        let mut address_iter = try_address.to_socket_addrs()?;
+
+        let listening_address = address_iter
+            .next()
+            .ok_or_else(|| format_err!("Could not parse address {}", try_address))?;
+
+        let host = host.unwrap_or("https://example.com").to_string();
+
+        let parsed_url = match Url::parse(&host) {
+            Ok(host) => Ok(host),
+            Err(_) => Url::parse(&format!("https://{}", host)),
+        }?;
+
+        let scheme = parsed_url.scheme();
+        if scheme != "http" && scheme != "https" {
+            failure::bail!("Your host scheme must be either http or https")
+        }
+        let is_https = scheme == "https";
+
+        let host = parsed_url.host_str().ok_or(format_err!("Invalid host, accepted formats are example.com, http://example.com, or https://example.com"))?.to_string();
```

Comment:
I suggest storing this as a `Url` rather than as a `String`, and adding an accessor method for `host_string` and one for `is_https` rather than storing these two specialized values directly on the struct.